### ISSUE TITLE
TypeError: fragments is undefined

### ISSUE THREAD

**Issue by margaretmoser** [apollostack/apollo-client#488]
I've been trying (and trying and trying) to get a mutation+updateQuery combination to work, and I'm encountering unexpected issues. I've been following the example provided in the apollo-client doc ("Updating query results with updateQueries").

This is a React application where a logged-in user can save and retrieve bookmarks, which are simple objects with a few scalar properties. I have a BookmarkMenu component that shows the user's bookmarks as a list. A separate component (GraphViewer) has a button allowing them to save a new one. The query for the list goes through fine, but the mutation throws 'TypeError: fragments is undefined' and then every further action gets 'Interrupted by an error up the chain'.

I've put the various parts of my code below. I've read through all the doc and issues I can find, and I've tried everything I can think of, including providing empty objects/arrays in the `fragments` param. But nothing has worked. I am not even sure why fragments are getting involved here (they're apparently not needed in the documentation example I'm following), so maybe that's where I'm going wrong. Can you help?

The gory details (error message is at the bottom):
My setup is based on react-apollo-starter-kit. Here's the client instantiation:
```
const ApolloApiClient = new ApolloClient({
  networkInterface,
  shouldBatch: true,
  dataIdFromObject: (obj) => obj.uuid
});
```
Here is the query I'm using in the bookmark menu:
```
    bookmarks:
    {
        query: graphql[tick]
          query getUserBookmarks($userId: String!) {
            user (userId:$userId) {
              uuid
              bookmarks {
                uuid
                [... other properties]
              }
            }
          }
        [tick],
        variables: {
          userId: state.auth.user.userId
        },
        forceFetch: false,
        returnPartialData: true,
      }
```
When I submit this query in my GraphiQL server, it returns this:
```
{
  "data": {
    "user": {
      "uuid": "ba88e490-4e07-...",
      "bookmarks": [
        { "uuid": "933d9670-5761-...", [other properties] },
        { "uuid": "964337c0-56d1-...", [other properties] },
        [...more bookmarks]
      ]}}}
```
This displays correctly in my React app (i.e. the list renders). The store shows:
```
apollo:
  data:
    [an entry for each bookmark]
    "ba88e490-4e07-...": {
      "uuid": "ba88e490-4e07-..."
      "bookmarks" [an entry for each bookmark]
    }
```
So far so good (I think). Now I want to add a bookmark and have it appear in the UI menu, preferably with optimistic rendering but definitely upon server response. Here's my mutation:
```
saveBookmark: ([bookmark properties]) => ({
  mutation: graphql[tick]
    mutation createBookmark(
      $createdBy: String!
      [...other bookmark properties]
    ) {
      createBookmark(
        createdBy: $createdBy
        [...other bookmark properties]
      ) {
        uuid
        [...other bookmark properties]
      }
    }
  [tick],
  variables: {
    createdBy: state.auth.user.userId,
    [...other properties]
  },
  updateQueries: {
    getUserBookmarks: (previousQueryResult, { mutationResult }) => {
        const bookmarks = [...previousQueryResult.user.bookmarks,
                            mutationResult.data.createBookmark];
        return { uuid: "ba88e490-4e07-...", bookmarks: bookmarks };
    }
  },
  optimisticResponse: { [...other properties], uuid: '' }
})
```
This is what GraphiQL returns for that mutation:
```
{
  "data": {
    "createBookmark": {
      "uuid": "55aea0c0-5765-...",
      [...other properties]
    }
  }
}
```
When I start the add from the UI, I can see from console logging that the mutation returns this:
```
{ 
  uuid: "ba88e490-4e07-..."
  bookmarks: [
    { uuid: "130dfa70-5767-...",  [...other properties]},
    [...more bookmarks...]
    { uuid: "55aea0c0-5765-...", [...other properties]}      <- the one I just created
] }
```
In Redux devtools, I see this:
```
APOLLO_MUTATION_INIT
action:
  mutationString:[exactly as above]
  mutationId:"2"
  type:"QUERY_RESULT"
  optimisticResponse: { [exactly as above] }
  [...other properties...]
  resultBehaviors/0/newResult: [exactly what's in the console log]
  queryFragments:undefined
TypeError: fragments is undefined

APOLLO_MUTATION_RESULT
action: [...most of the 'action' property as above]
Interrupted by an error up the chain
```
Here's the console stack trace:
```
createFragmentMap@[apollo-client/queries/getFromAST.js]
mutationResultQueryResultReducer@[apollo-client/data/mutationResults.js]
data/<@[apollo-client/data/store.js, line 65]
data@[apollo-client/data/store.js, line 55]
optimistic@[apollo-client/optimistic-data/store.js, line 19]
apolloReducer@[apollo-client/data/store.js, line 27]
...
```
I've tried commenting out my optimisticResponse and I get the same result, erroring at the same line.

**Comment by HriBB** [comment]
Could be related to #470

**Comment by stretchkennedy** [comment]
This looks like it's related to #470 and #475, but I was able to circumvent that by writing something similar to 

```jsx
export default connect({
  mapQueriesToProps: () => ({
    data: {
      query: gql`
        query foos {
          foos {
            id
          }
        }
      `,
      fragments: [], // fix for https://github.com/apollostack/apollo-client/pull/475
    },
  }),
  mapMutationsToProps: () => ({
    deleteFoos: (id) => {
      return {
        mutation: gql`
          mutation DeleteFooMutation($input: DeleteFooMutationInput!) {
            deleteFoo(input: $input) {
              fooId
            }
          }
        `,
        variables: {
          input: {
            id,
          },
        },
        updateQueries: {
          foos: (previousQueryResult, { mutationResult }) => {
            const newResult = cloneDeep(previousQueryResult);
            const newFoos = previousQueryResult.foos.filter(
              foo => foo.id !== mutationResult.data.deleteFoo.fooId
            );
            newResult.foos = newFoos;
            return newResult;
          },
        },
        optimisticResponse: {
          deleteFoo: {
            fooId: id,
          },
        },
      };
    },
  }),
})(ManageFoosComponent);
```

The empty array of fragments had to be provided _to the query_, not to the mutation.

**Comment by margaretmoser** [comment]
I'll give that a shot, thank you.

**Comment by margaretmoser** [comment]
I tried it, and indeed it did work. I just started working on this 3 days ago... and #470 was closed 4 days ago but hasn't made it to the npm package yet. Looks like I should switch to pulling master instead of relying on the package since this is moving so rapidly.

Anyway, thanks very much – it's so satisfying to get the round trip working! I've now made several bookmarks just to see them appear in the menu...
